planb/PBPackageInstaller.m

```diff
@@ -42,30 +37,76 @@ @interface PBPackageInstaller ()
 
 @implementation PBPackageInstaller
 
-- (instancetype)initWithReceiptName:(NSString *)receiptName
-                        packagePath:(NSString *)packagePath
-                       targetVolume:(NSString *)targetVolume {
+static NSString *const kHdiutilPath = @"/usr/bin/hdiutil";
+static NSString *const kInstallerPath = @"/usr/sbin/installer";
+static NSString *const kPkgutilPath = @"/usr/sbin/pkgutil";
+
+- (instancetype)initWithURL:(NSURL *)packageURL
+                receiptName:(NSString *)receipt {
   self = [super init];
 
   if (self) {
-    if (!receiptName.length || !packagePath.length || !targetVolume.length) {
+    if (!receipt.length || !packageURL.absoluteString.length) {
       return nil;
     }
 
-    _receiptName = [receiptName copy];
-    _packagePath = [packagePath copy];
-    _targetVolume = [targetVolume copy];
+    _receiptName = [receipt copy];
+    _packageURL = [packageURL copy];
+    _mountPoint = [self generateMountPoint];
+
+    _downloadAttemptsMax = 5;
+    _downloadTimeoutSeconds = 300;
+    _session = [NSURLSession sharedSession];
   }
 
   return self;
 }
 
+- (void)dealloc {
+  if (self.mountPoint) {
+    NSError *err;
+    [[NSFileManager defaultManager] removeItemAtPath:self.mountPoint error:&err];
+    if (err) {
+      PBLog(@"%@ Error: could not delete temporary mount point %@: %@",
+            self.logPrefix, self.mountPoint, err.localizedDescription);
+    }
+  }
+}
+
+- (BOOL)install {
+  NSString *path = [self downloadPackage];
+  if (!path) return NO;
+
+  BOOL success = [self diskImageAttach:path];
+  if (!success) return NO;
+
+  PBLog(@"%@ Mounted at %@", self.logPrefix, self.mountPoint);
+
+  // We don't care if this works or not.
+  [self forgetPackageWithName:self.receiptName];
+
+  // TODO(rah): Add ability to install to a different volume.
+  success &= [self installPackageFromLocation:[self firstPackageOnImage]
```

Comment:
Why the need to `and` here? If diskImageAttach fails the method returns. So at this point success is always `1`.